### PR TITLE
Allow getOption[Label|Value] props in Typeahead

### LINKED ISSUE #1242: TypeAhead kit not recognizing getOptionLabel and getOptionValue props
**Describe the bug**
When implementing react TypeAhead we will often need to customize the `label and value` to be `name and id`. React select has a prop `getOptionLabel` and `getOptionValue` that we could use for this matter but it seems that these two props are not working on playbook TypeAhead. The customization of label and value is something that we will need quite often so it would be great if we could find a way to make `getOptionLabel` and `getOptionValue` props useable with playbook. I've been playing around with playbook code to see why those props are not working but no luck yet. Right now we have a workaround on the Nitro side but it is not ideal. 
Thank you!
```
/* @flow */

import React from 'react'
import { Typeahead } from '../..'

const options = [
  { name: 'Windows', id: '#FFA500' },
  { name: 'Siding', id: '#FF0000' },
  { name: 'Doors', id: '#00FF00' },
  { name: 'Roofs', id: '#0000FF' },
]

const TypeaheadExample = () => {
  return (
    <>
      <Typeahead
          isMulti
          label="Example"
          getOptionLabel={(option) => option.name}
          getOptionValue={(option) => option.id}
          options={options}
          placeholder=""
      />
    </>
  )
}

export default TypeaheadExample
```

Screen Shot 2020-12-09 at 10 57 33 AM

**Expected behavior**
Being able to pass `getOptionLabel` and `getOptionValue` to TypeAhead kit
https://medium.com/@lahiru0561/react-select-with-custom-label-and-custom-search-122bfe06b6d7

### PR DESCRIPTION
**Pull request by thestephenmarshall**
Fixes #1242 

#### Screens

N/A

#### Breaking Changes

none

#### Runway Ticket URL

https://nitro.powerhrg.com/runway/backlog_items/NUXE-342

#### How to test this

See https://playbook.powerapp.cloud/kits/typeahead examples

#### Checklist:

- [x] **LABELS** Add a label: `enhancement`, `bug`, `improvement`, `new kit`, `deprecated`, or `breaking`. See [Changelog & Labels](https://github.com/powerhome/playbook/wiki/Changelog-&-Labels) for details.
- [x] **URGENCY** Please select a release milestone
- [x] **DEPLOY** Please add the `Milano` label when you are ready for a review.
- [x] **SCREENSHOT** Please add a screen shot or two.
- [x] **SPECS** Please cover your changes with specs.
- [x] **READ DOCS** Please make sure you have read and understand the [Playbook Release Process](https://github.com/powerhome/playbook/wiki/Playbook-Releases)

*The normal release cut off deadline is 3p EDT each week. Please reach out to the release team if you have an urgent request or need a release off cycle.*

### CODE REVIEW

File: playbook/app/pb_kits/playbook/pb_typeahead/_typeahead.jsx [hunk 1]
```diff
@@ -27,6 +27,8 @@ type Props = {
   async?: boolean,
   label?: string,
   loadOptions?: noop | string,
+  getOptionLabel?: () => any,
+  getOptionValue?: () => any,
```
Comment (thestephenmarshall) [inline playbook/app/pb_kits/playbook/pb_typeahead/_typeahead.jsx:31]: First form of documentation starts with the type

File: playbook/app/pb_kits/playbook/pb_typeahead/_typeahead.jsx [hunk 2]
```diff
@@ -50,6 +51,7 @@ const Typeahead = (props: Props) => {
       Placeholder,
       ValueContainer,
     },
+    defaultOptions: true,
```
Comment (thestephenmarshall) [inline playbook/app/pb_kits/playbook/pb_typeahead/_typeahead.jsx:54]: Alpha ordering

File: playbook/app/pb_kits/playbook/pb_typeahead/_typeahead.jsx [hunk 3]
```diff
@@ -58,6 +60,8 @@ const Typeahead = (props: Props) => {
   }
 
   if (typeof(props.loadOptions) === 'string') selectProps.loadOptions = get(window, props.loadOptions)
+  if (typeof(props.getOptionLabel) === 'string') selectProps.getOptionLabel = get(window, props.getOptionLabel)
+  if (typeof(props.getOptionValue) === 'string') selectProps.getOptionValue = get(window, props.getOptionValue)
```
Comment (thestephenmarshall) [inline playbook/app/pb_kits/playbook/pb_typeahead/_typeahead.jsx:64]: Just like `loadOptions`, we need to check for the presence of the prop. If this pattern emerges in another kit, we should abstract this away to a utility method for re-use.

File: playbook/app/pb_kits/playbook/pb_typeahead/components/Option.jsx [hunk 4]
```diff
@@ -21,12 +20,12 @@ const Option = (props: any) => {
             <User
                 align="left"
                 avatarUrl={imageUrl}
-                name={label}
+                name={props.label}
```
Comment (thestephenmarshall) [inline playbook/app/pb_kits/playbook/pb_typeahead/components/Option.jsx:23]: The `data.label` isn't going to be correct if you use `getOptionLabel`, so just use `props.label` instead.

File: playbook/app/pb_kits/playbook/pb_typeahead/docs/_typeahead_with_pills_async.html.erb [hunk 5]
```diff
@@ -1,4 +1,13 @@
...
+<%= pb_rails("typeahead", props: {
+  async: true,
+  get_option_label: 'getOptionLabel',
+  get_option_value: 'getOptionValue',
+  load_options: 'asyncPillsPromiseOptions',
+  label: "Github Users",
+  name: :foo,
+  pills: true,
+  placeholder: "type the name of a Github user"
+}) %>
```
Comment (thestephenmarshall) [inline playbook/app/pb_kits/playbook/pb_typeahead/docs/_typeahead_with_pills_async.html.erb:10]: Was unreadable.

File: playbook/app/pb_kits/playbook/pb_typeahead/docs/_typeahead_with_pills_async.html.erb [hunk 6]
```diff
@@ -27,4 +36,11 @@
       }
     })
   }
+
+  window.getOptionLabel = function(option) {
+    return option.name;
+  }
+  window.getOptionValue = function(option) {
+    return option.id;
+  }
```
Comment (thestephenmarshall) [inline playbook/app/pb_kits/playbook/pb_typeahead/docs/_typeahead_with_pills_async.html.erb:45]: From the Rails plain JS side, here is how you provide the function each prop is expecting.

File: playbook/app/pb_kits/playbook/pb_typeahead/typeahead.rb [hunk 7]
```diff
@@ -45,6 +47,9 @@ def typeahead_with_pills_options
           placeholder: placeholder
         }
 
+        base_options.merge!({getOptionLabel: get_option_label}) if get_option_label.present?
+        base_options.merge!({getOptionValue: get_option_value}) if get_option_value.present?
+
```
Comment (thestephenmarshall) [inline playbook/app/pb_kits/playbook/pb_typeahead/typeahead.rb:52]: It's silly to send `nil` for the prop which will convert to `null` which isn't handled by `react-select` lib.

**Comment by thestephenmarshall** [comment]
Fixes #1244

**Review by bh247484 — APPROVED** [review]

**Pull request merged**